src/main/java/entities/Player.java

```diff
@@ -85,12 +86,30 @@ public Card playCard(int id) {
         return null;
     }
 
+    public boolean spendEssence(int num) {
+        if (num > getNumOfEssence()) {
+            return false;
+        }
+
+        Iterator<Card> iterator = hand.iterator();
+        while (iterator.hasNext() && num > 0) {
+            Card card = iterator.next();
+            if (card instanceof EssenceCard) {
+                iterator.remove();
+                num--;
+            }
+        }
+
+        return true;
+    }
+
```

Comment:
I see why we had to use an iterator here, it's because if we do the for loop approach, we are removing elements in the list as we go, and that will mess up the iteration. But also we don't really have to use an iterator. The more easy-to-read solution is to create a list and collect the cards that needs to be removed, then use the removeAll() method at the very end.

Check the suggestion below this comment please!